source/server/listener_manager_impl.cc

```diff
@@ -843,14 +867,29 @@ std::unique_ptr<Network::FilterChain> ListenerFilterChainFactoryBuilder::buildFi
   ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
       transport_socket, parent_.messageValidationVisitor(), config_factory);
 
-  std::vector<std::string> server_names(filter_chain.filter_chain_match().server_names().begin(),
-                                        filter_chain.filter_chain_match().server_names().end());
-
-  return std::make_unique<FilterChainImpl>(
+  return std::make_unique<TcpFilterChainImpl>(
       config_factory.createTransportSocketFactory(*message, factory_context_,
                                                   std::move(server_names)),
       parent_.parent_.factory_.createNetworkFilterFactoryList(filter_chain.filters(), parent_));
 }
 
+QuicListenerFilterChainFactoryBuilder::QuicListenerFilterChainFactoryBuilder(
+    ListenerImpl& listener,
+    Server::Configuration::TransportSocketFactoryContextImpl& factory_context)
+    : ListenerFilterChainFactoryBuilder(listener, factory_context) {}
+
+std::unique_ptr<Network::FilterChain> QuicListenerFilterChainFactoryBuilder::buildFilterChain(
+    const ::envoy::api::v2::listener::FilterChain& filter_chain) const {
+  // Initialize filter chain for QUIC.
+  if (!filter_chain.has_tls_context()) {
+    return nullptr;
+  }
+  auto& tls_context_factory =
+      Config::Utility::getAndCheckFactory<Ssl::ContextConfigFactory>("downstream_tls_context");
```

Comment:
"downstream_tls_context" should be a constant somewhere. 

Can you talk a bit more about why we need to load the factory on this way vs. just directly instantiating it? It's a little confusing. I suppose this is due to pluggable TLS implementations?

Reply:
Yes, all the registered class work in this PR is just to make envoy core code not depends on TLS extensions.